edb/server/pgcon/rust_transport.py

```diff
@@ -0,0 +1,475 @@
+                        socket.TCP_KEEPIDLE, TCP_KEEPIDLE)
+    if hasattr(socket, 'TCP_KEEPALIVE'):
+        sock.setsockopt(socket.IPPROTO_TCP,
+                        socket.TCP_KEEPALIVE, TCP_KEEPIDLE)
+
+    # TCP_KEEPINTVL: The time (in seconds) between individual keepalive probes.
+    if hasattr(socket, 'TCP_KEEPINTVL'):
+        sock.setsockopt(socket.IPPROTO_TCP,
+                        socket.TCP_KEEPINTVL, TCP_KEEPINTVL)
+
+    # TCP_KEEPCNT: The maximum number of keepalive probes TCP should send
+    # before dropping the connection.
+    if hasattr(socket, 'TCP_KEEPCNT'):
+        sock.setsockopt(socket.IPPROTO_TCP,
+                        socket.TCP_KEEPCNT, TCP_KEEPCNT)
+
+
+P = TypeVar('P', bound=asyncio.Protocol)
+
+
+async def create_postgres_connection(
+    dsn: str | ConnectionParams,
+    protocol_factory: Callable[[], P],
+    *,
+    source_description: Optional[str] = None,
+    state_change_callback: Optional[StateChangeCallback] = None
+) -> Tuple[PGRawConn, P]:
+    """
+    Open a PostgreSQL connection to the address specified by the DSN or
+    ConnectionParams, creating the user protocol from the protocol_factory.
+
+    This method establishes the connection asynchronously. When successful, it
+    returns a (PGRawConn, protocol) pair.
+    """
+    if isinstance(dsn, str):
+        dsn = ConnectionParams(dsn=dsn)
+    connect_timeout = dsn.connect_timeout
+    try:
+        state = PyConnectionState(
+            dsn._params,
+            "postgres",
+            str(get_pg_home_directory())
+        )
+    except Exception as e:
+        raise ValueError(e)
+    ready_future: asyncio.Future = asyncio.Future()
+    pg_state = PGState(
+        parameters={},
+        cancellation_key=None,
+        auth=None,
+        server_error=None,
+        ssl=False
+    )
+
+    user_protocol = protocol_factory()
+    protocol, host, port, initial_transport = await _create_connection(
+        lambda: PGConnectionProtocol(
+            state,
+            user_protocol,
+            ready_future,
+            pg_state
+        ),
+        connect_timeout,
+        state.config.host_candidates
+    )
+
+    upgraded_transport = None
+    try:
+        update = RustTransportUpdate(
+            state,
+            initial_transport,
+            state_change_callback,
+            pg_state,
+            host if protocol != "unix" else None,
+            ready_future
+        )
+        state.update = update
+        state.drive_initial()
+
+        await ready_future
+        upgraded_transport = update.raw_transport
+        conn = PGRawConn(
+            source_description,
+            ConnectionParams._create(state.config),
+            upgraded_transport,
+            pg_state,
+            (host, port)
+        )
+        upgraded_transport.set_protocol(user_protocol)
+
+        # Notify the user protocol of successful connection
+        user_protocol.connection_made(conn)
```

Comment:
How do we make sure that `connection_made()` is always called before any `data_received()` calls? Given the concerns of `ready_future` being set multiple times, I'm thinking if it's possible to combine this logic into the `PGConnectionProtocol` in e.g. the `connection_made()` callback or something, and lift some - if not all - usages of `ready_future`.

Reply:
Everything now happens in PGConnectionProtocol which should hopefully make this more obviously correct.